### PR TITLE
Sync XSD "IntelliSense" with VS repo automatically on insertion

### PR DESCRIPTION
**Pull request by rainersigwald**
We've been manually managing this and naturally forgotten it a few times. That's not cool. Let's let a robot do it.

This requires change to the releases: https://dev.azure.com/devdiv/DevDiv/_releaseProgress?releaseId=749728&_a=release-history but it will be harmless to do this first and then change them.

Example internal PR: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/262150?_a=files

### CODE REVIEW

File: src/MSBuild/MSBuild/Microsoft.Build.CommonTypes.xsd [hunk 1]
```diff
@@ -5624,7 +5624,7 @@ elementFormDefault="qualified">
             </xs:complexContent>
         </xs:complexType>
     </xs:element>
-
+<!-- a change made on GitHub that should flow -->
```
Comment (rainersigwald) [inline src/MSBuild/MSBuild/Microsoft.Build.CommonTypes.xsd:5627]: I should of course revert this proof that it worked.

File: src/MSBuild/Update-MSBuildXsds.ps1 [hunk 2]
```diff
@@ -0,0 +1,9 @@
+$ErrorActionPreference = 'Stop'
+
+cmd /c set
```
Comment (rainersigwald) [inline src/MSBuild/Update-MSBuildXsds.ps1:3]: remove

File: src/MSBuild/MSBuild.csproj [hunk 3]
```diff
@@ -247,6 +240,14 @@
     <ProjectReference Include="..\Tasks\Microsoft.Build.Tasks.csproj" />
   </ItemGroup>
 
+  <Target Name="CopyXsds"
+          BeforeTargets="AfterBuild">
+    <Copy SourceFiles="@(XsdsForVS)"
+          DestinationFiles="@(XsdsForVS->'$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'xsd'))%(RecursiveDir)%(Filename)%(Extension)')"
```
Comment (Forgind) [inline src/MSBuild/MSBuild.csproj:246]: Why does this work? I thought you didn't assign it metadata.

File: src/MSBuild/MSBuild.csproj [hunk 4]
```diff
@@ -216,16 +216,9 @@
...
-      <SubType>Designer</SubType>
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="MSBuild\Microsoft.Build.Core.xsd">
-      <SubType>Designer</SubType>
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="Microsoft.Build.xsd">
-      <SubType>Designer</SubType>
+    <XsdsForVS Include="**\*.xsd" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" />
```
Comment (Forgind) [inline src/MSBuild/MSBuild.csproj:219]: This hits extra xsds and sometimes multiple copies of one xsd, right? I don't think that's a problem, although it might hit performance slightly, but I wanted to check if I understood.

File: src/MSBuild/MSBuild.csproj (hunk 3 above)
Comment (rainersigwald) [reply src/MSBuild/MSBuild.csproj:246]: I don't understand the question. Can you rephrase? Are you asking about the [well-known metadata](https://docs.microsoft.com/visualstudio/msbuild/msbuild-well-known-item-metadata?view=vs-2019)?

File: src/MSBuild/MSBuild.csproj (hunk 4 above)
Comment (rainersigwald) [reply src/MSBuild/MSBuild.csproj:219]: I don't understand the question. Can you rephrase? Specifically what do you mean by "extra xsds" and by "multiple copies of one xsd"?

Comment (Forgind) [reply src/MSBuild/MSBuild.csproj:219]: I was thinking that if it's looking for anything.xsd, it would also find RuleSet.xsd, CustomDictionary.xsd, etc. and a different copy for each build configuration, but it occurred to me that that's all under the artifacts folder, so I think I was wrong.

File: src/MSBuild/MSBuild.csproj (hunk 3 above)
Comment (Forgind) [reply src/MSBuild/MSBuild.csproj:246]: I'd forgotten about those. Thanks!

File: src/MSBuild/MSBuild.csproj (hunk 4 above)
Comment (BenVillalobos) [inline src/MSBuild/MSBuild.csproj:219]: There are two "Default exclude" items here,  so presumably "DefaultItemExcludes" doesn't include the "DefaultExcludesInProjectFolder". How does one know or find out how many default exclude items there are?

Comment (Forgind) [reply src/MSBuild/MSBuild.csproj:219]: There's probably a better way, but I build a project that excludes those, catch it in a debugger, break on the code that handles that in `Project.cs`, and inspect what it gets turned into.

Comment (rainersigwald) [reply src/MSBuild/MSBuild.csproj:219]: I'm following the SDK's approaches here:

https://github.com/dotnet/sdk/blob/e053f40806bf1e50b01271cbddc57759d24f0b0b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets#L22-L42

https://github.com/dotnet/sdk/blob/062d9a036ea335833e472d3b3c17174ba62c5c19/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.props#L37

No real way to know other than looking for analogies there that I know of.
